include/envoy/grpc/async_client.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <chrono>
+
+#include "envoy/common/optional.h"
+#include "envoy/common/pure.h"
+#include "envoy/grpc/status.h"
+#include "envoy/http/header_map.h"
+
+#include "google/protobuf/descriptor.h"
+
+namespace Envoy {
+namespace Grpc {
+
+/**
+ * An in-flight gRPC stream.
+ */
+template <class RequestType> class AsyncClientStream {
+public:
+  virtual ~AsyncClientStream() {}
+
+  /**
+   * Send request message to the stream.
+   * @param request protobuf serializable message.
+   */
+  virtual void sendMessage(const RequestType& request) PURE;
```

Comment:
This is not possible in the Envoy filter model, we don't know when the message was actually sent.